packages/core/src/mesh/BlendShapeManager.ts

```diff
@@ -0,0 +1,330 @@
+      textureWidth = maxTextureSize;
+    }
+
+    const shapeCount = this._blendShapes.length;
+    let blendShapeDataTexture = this._dataTexture;
+    const blendShapeCount = this._blendShapes.length;
+
+    blendShapeDataTexture && blendShapeDataTexture.destroy();
+
+    blendShapeDataTexture = new Texture2DArray(
+      this._engine,
+      textureWidth,
+      textureHeight,
+      blendShapeCount,
+      TextureFormat.R32G32B32A32,
+      false
+    );
+    blendShapeDataTexture.filterMode = TextureFilterMode.Point;
+
+    this._dataTextureBuffer = new Float32Array(shapeCount * textureWidth * textureHeight * 4);
+    this._dataTexture = blendShapeDataTexture;
+    this._dataTextureInfo.setValue(vertexPixelStride, textureWidth, textureHeight);
+  }
+
+  private _updateDataToTexture(vertexCount: number, force: boolean): void {
+    const {
+      _blendShapes: blendShapes,
+      _dataTexture: dataTexture,
+      _dataTextureBuffer: buffer,
+      _subDataDirtyFlags: subDataDirtyFlags
+    } = this;
+
+    let offset = 0;
+    for (let i = 0, n = blendShapes.length; i < n; i++) {
+      const subDirtyFlag = subDataDirtyFlags[i];
+      const subBlendShapeDataStride = dataTexture.width * dataTexture.height * 4;
+      if (force || subDirtyFlag.flag) {
+        const { frames } = blendShapes[i];
+        const frameCount = frames.length;
+        const endFrame = frames[frameCount - 1];
+        if (frameCount > 0 && endFrame.deltaPositions.length !== vertexCount) {
+          throw "BlendShape frame deltaPositions length must same with mesh vertexCount.";
+        }
+        const { deltaPositions, deltaNormals, deltaTangents } = endFrame;
+        offset = i * subBlendShapeDataStride;
+        for (let j = 0; j < vertexCount; j++) {
+          const position = deltaPositions[j];
+          buffer[offset] = position.x;
+          buffer[offset + 1] = position.y;
+          buffer[offset + 2] = position.z;
+          offset += 4;
+
+          if (deltaNormals) {
+            const normal = deltaNormals[j];
+            buffer[offset] = normal.x;
+            buffer[offset + 1] = normal.y;
+            buffer[offset + 2] = normal.z;
+            offset += 4;
+          }
+
+          if (deltaTangents) {
+            const tangent = deltaTangents[j];
+            buffer[offset] = tangent.x;
+            buffer[offset + 1] = tangent.y;
+            buffer[offset + 2] = tangent.z;
+            offset += 4;
+          }
+        }
+        subDirtyFlag.flag = false;
+      }
+    }
+    dataTexture.setPixelBuffer(0, buffer);
+  }
+
+  /**
+   * @internal
+   */
+  _releaseMemoryCache(): void {
+    this._layoutDirtyFlag.destroy();
+    const dataChangedFlags = this._subDataDirtyFlags;
+    for (let i = 0, n = dataChangedFlags.length; i < n; i++) {
+      dataChangedFlags[i].destroy();
+    }
+
+    this._layoutDirtyFlag = null;
+    this._subDataDirtyFlags = null;
+    this._blendShapes = null;
```

Comment:
how can i get blendshape name if you release it

Reply:
done